src/rgw/driver/rados/topic_migration.cc

```diff
@@ -0,0 +1,217 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright contributors to the Ceph project
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ *
+ */
+
+#include "topic_migration.h"
+#include "services/svc_zone.h"
+#include "rgw_sal_rados.h"
+
+namespace rgwrados::topic_migration {
+
+namespace {
+
+// migrate v1 notification metadata for a single bucket
+int migrate_notification(const DoutPrefixProvider* dpp, optional_yield y,
+                         rgw::sal::RadosStore* driver, const rgw_raw_obj& obj)
+{
+  // parse bucket name and marker of out "pubsub.{tenant}.bucket.{name}/{marker}"
+  std::string tenant;
+  std::string bucket_name;
+  std::string marker;
+  int r = rgw::sal::RadosBucket::deconstruct_topics_oid(obj.oid, tenant, bucket_name, marker, dpp);
+  if (r < 0) {
+    ldpp_dout(dpp, 1) << "failed to read tenant, bucket name and marker from " << obj
+                      << " with: " << cpp_strerror(r)
+                      << " expected format pubsub.{tenant}.bucket.{name}/{marker}!" << dendl;
+    return r;
+  }
+
+  // migrate the notifications
+  rgw_pubsub_bucket_topics bucket_topics;
+  rgw_bucket rgw_bucket_info(tenant, bucket_name);
+  rgw::sal::RadosBucket rados_bucket(driver, rgw_bucket_info);
+  RGWObjVersionTracker bucket_topics_objv;
+  r = rados_bucket.read_topics(bucket_topics, &bucket_topics_objv, y, dpp);
+  if (r == -ENOENT) {
+    return 0; // ok, someone else already migrated
+  }
+  if (r < 0) {
+    ldpp_dout(dpp, 1) << "failed to read v1 bucket notifications from " << obj
+                      << " with: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  // in a for-loop that retries ECANCELED errors:
+  // {
+  // load the corresponding bucket by name
+  // break if marker doesn't match loaded bucket's
+  // break if RGW_ATTR_BUCKET_NOTIFICATION xattr already exists
+  // write RGW_ATTR_BUCKET_NOTIFICATION xattr
+  // }
+  std::unique_ptr<rgw::sal::Bucket> bucket;
+  do {
+    r = driver->load_bucket(dpp, rgw_bucket_info, &bucket, y);
+    if (r < 0) {
+      ldpp_dout(dpp, 1) << "failed to load the bucket from " << obj
+                        << " with: " << cpp_strerror(r) << dendl;
+      continue;
+    }
+
+    if (bucket->get_marker() != marker && bucket->get_attrs().contains(RGW_ATTR_BUCKET_NOTIFICATION)) {
+      break;
+    }
+
+    bufferlist bl;
+    bucket_topics.encode(bl);
+    bucket->get_attrs()[RGW_ATTR_BUCKET_NOTIFICATION] = std::move(bl);
+  } while (r != ECANCELED);
+
+  // delete v1 notification obj with Bucket::remove_topics()
+  if (bucket) {
+    r = bucket->remove_topics(&bucket_topics_objv, y, dpp);
```

Comment:
@AliMasarweh to expand on this:

the bucket's `marker` only changes if the bucket gets deleted and recreated. so if we load the bucket and find a different marker, that means the notification referred to an older version of the bucket, so the notification is stale and should be deleted without trying to migrate it

if we call `bucket->remove_topics()` in that case, it would try to delete an object "pubsub.{tenant}.bucket.{name}/{marker}" with the wrong marker instead of the `const rgw_raw_obj& obj` that we passed in. calling `rados_bucket.remove_topics()` instead should do the right thing, since we constructed it with the correct marker

Reply:
so we should load the bucket again to ensure that it has the correct marker?